stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+        LS_backward_ = SUNDenseLinearSolver(nv_state_backward_, A_backward_);
+      }
+      cvodes_mem_ = CVodeCreate(solver_forward);
+
+      if (cvodes_mem_ == nullptr) {
+        throw std::runtime_error("CVodeCreate failed to allocate memory");
+      }
+    }
+
+    virtual ~cvodes_solver() {
+      SUNMatDestroy(A_forward_);
+      SUNMatDestroy(A_backward_);
+      if (N_ > 0) {
+        SUNLinSolFree(LS_forward_);
+        SUNLinSolFree(LS_backward_);
+      }
+      N_VDestroy_Serial(nv_state_forward_);
+      N_VDestroy_Serial(nv_state_backward_);
+      N_VDestroy_Serial(nv_quad_);
+      N_VDestroy_Serial(nv_absolute_tolerance_forward_);
+      N_VDestroy_Serial(nv_absolute_tolerance_backward_);
+
+      if (cvodes_mem_) {
+        CVodeFree(&cvodes_mem_);
+      }
+    }
+  };
+
+ public:
+  /**
+   * Construct cvodes_integrator object. Note: All arguments must be stored as
+   * copies if in doubt. The reason is that the references can go out of scope,
+   * since the work done from the integrator is in the chain method.
+   *
+   * @param function_name Calling function name (for printing debugging
+   * messages)
+   * @param f Right hand side of the ODE
+   * @param y0 Initial state
+   * @param t0 Initial time
+   * @param ts Times at which to solve the ODE at. All values must be sorted and
+   *   not less than t0.
+   * @param relative_tolerance_forward Relative tolerance for forward problem
+   * passed to CVODES
+   * @param absolute_tolerance_forward Absolute tolerance for forward problem
+   * passed to CVODES
+   * @param relative_tolerance_backward Relative tolerance for backward problem
+   * passed to CVODES
+   * @param absolute_tolerance_backward Absolute tolerance for backward problem
+   * passed to CVODES
+   * @param relative_tolerance_quadrature Relative tolerance for quadrature
+   * problem passed to CVODES
+   * @param absolute_tolerance_quadrature Absolute tolerance for quadrature
+   * problem passed to CVODES
+   * @param max_num_steps Upper limit on the number of integration steps to
+   *   take between each output (error if exceeded)
+   * @param num_steps_between_checkpoints Number of integrator steps after which
+   * a checkpoint is stored for the backward pass
+   * @param interpolation_polynomial type of polynomial used for interpolation
+   * @param solver_forward solver used for forward pass
+   * @param solver_backward solver used for backward pass
+   *   take between each output (error if exceeded)
+   * @param[in, out] msgs the print stream for warning messages
+   * @param args Extra arguments passed unmodified through to ODE right hand
+   * side function
+   * @return Solution to ODE at times \p ts
+   * @return a vector of states, each state being a vector of the
+   * same size as the state variable, corresponding to a time in ts.
+   */
+  template <typename FF, require_eigen_col_vector_t<T_y0>* = nullptr>
+  cvodes_integrator_adjoint_vari(
```

Comment:
imo it would be nice to have this moved up nearer to where stuff is defined like 

```cpp
cvodes_integrator_adjoint {
the_member_objects;
public:
 constructor()
private:
  internal_stuff;
};
```

it's just a bit easier to sort out what's going on when I read the class definition

Reply:
ok